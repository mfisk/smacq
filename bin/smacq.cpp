#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <string.h>
#include <smacq.h>
#include <SmacqGraph.h>
#include <SmacqScheduler.h>
#include "config.h"

#define MAX_QUERY_SIZE 4096*100

static struct smacq_options options[] = {
  {"t", {string_t:NULL}, "Describe the specified type", SMACQ_OPT_TYPE_STRING},
  {"m", {boolean_t:0}, "Multiple queries on STDIN", SMACQ_OPT_TYPE_BOOLEAN},
  {"f", {string_t:"-"}, "File to read queries from", SMACQ_OPT_TYPE_STRING},
  {"pregraph", {boolean_t:0}, "Show pre-initialization graph", SMACQ_OPT_TYPE_BOOLEAN},
  {"g", {boolean_t:0}, "Show final graph", SMACQ_OPT_TYPE_BOOLEAN},
  {"q", {boolean_t:0}, "Ignore warnings", SMACQ_OPT_TYPE_BOOLEAN},
  END_SMACQ_OPTIONS
};

#ifdef REFINFO

#include <signal.h>
RefMap REFS;

inline void print_refs(int x) {
        FILE * fh = fopen("/tmp/smacq.refs", "w");
        std::map<void*,std::map<void*, Counter> >::iterator i;
        std::map<void*, Counter>::iterator j;
        for (i=REFS.begin(); i != REFS.end(); ++i) {
           for (j=i->second.begin(); j != i->second.end(); ++j) {
                if (j->second.count) {
                   fprintf(fh, "%p\t%p\t%hd\n", i->first, j->first, j->second.count);
                }
           }
        }
        fclose(fh);
}
#endif

int smacqq(int argc, char ** argv) {
  smacq_opt multiple, optimize, qfile, showpregraph, showgraph, showtype, quiet;
  int qargc;
  char ** qargv;
  DTS dts;
  FILE * fh;
  SmacqGraphContainer * graphs = NULL;
  DtsObject product;

#ifdef REFINFO
  signal(SIGUSR1, print_refs);
#endif

  if (argc <= 1) {
	  fprintf(stderr, "Usage: %s [-m] query\n", argv[0]);
	  fprintf(stderr, "Version %s; Build date %s\n", 
		  PACKAGE_VERSION, SMACQ_BUILD_DATE);
	  return -1;
  }

  struct smacq_optval optvals[] = {
		  {"t", &showtype},
		  {"m", &multiple},
		  {"f", &qfile},
		  {"O", &optimize},
		  {"pregraph", &showpregraph},
		  {"g", &showgraph},
		  {"q", &quiet},
		  {NULL, NULL}
  };
  smacq_getoptsbyname(argc-1, argv+1, &qargc, &qargv, options, optvals);

  if (quiet.boolean_t) {
	dts.set_no_warnings();
  }

  if (showtype.string_t) {
	dts_typeid tid = dts.requiretype(showtype.string_t);
   	assert(tid);
   	dts_type * t = dts.type_bynum(tid);
   	std::vector<struct dts_field_info*>::iterator i;

   	printf("Type \"%s\" defines the following fields:\n", showtype.string_t);
   	for (i=t->fields.begin(); i != t->fields.end(); ++i) {
		if (*i) {
       		printf("%30s: type %s\n", (*i)->desc.name, (*i)->desc.type);
		}
   	}
   	exit(0);
  }

  SmacqScheduler s;

  if (multiple.boolean_t) {
      char * queryline;
      int qno=1;

      queryline = (char*)malloc(MAX_QUERY_SIZE);

      if (qargc) {
	      fprintf(stderr, "Cannot specify query on command line with -m option\n");
	      return -1;
      }

      if (!strcmp(qfile.string_t, "-")) {
	      fh = stdin;
      } else {
	      fh = fopen(qfile.string_t, "r");
      }

      while(fgets(queryline, MAX_QUERY_SIZE, fh)) {
	      /* Chomp newline */
	      if (queryline[strlen(queryline)-1] == '\n')
		      queryline[strlen(queryline)-1] = '\0';

	      SmacqGraphContainer * newgraph = SmacqGraph::newQuery(&dts, &s, 1, &queryline);
	      if (!newgraph) {
		      fprintf(stderr, "Fatal error at line %d\n", qno);
		      exit(-1);
	      }
      	      if (graphs) {
		graphs->add_graph(newgraph, true);
	      } else {
		graphs = newgraph;
	      }
	      qno++;
      }

  } else {
    graphs = SmacqGraph::newQuery(&dts, &s, qargc, qargv);
    if (!graphs) exit(-1);
  }

  if (showpregraph.boolean_t) {
      graphs->print(stderr, 8);
  }

  graphs->init(&dts, &s);

  if (showgraph.boolean_t) {
      graphs->print(stderr, 8);
  }

  s.seed_produce(graphs);
  return (! s.busy_loop());
}

