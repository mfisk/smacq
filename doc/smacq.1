.\" This -*- nroff -*- file has been generated from
.\" DocBook SGML with docbook-to-man on Debian GNU/Linux.
...\"
...\"	transcript compatibility for postscript use.
...\"
...\"	synopsis:  .P! <file.ps>
...\"
.de P!
\\&.
.fl			\" force out current output buffer
\\!%PB
\\!/showpage{}def
...\" the following is from Ken Flowers -- it prevents dictionary overflows
\\!/tempdict 200 dict def tempdict begin
.fl			\" prolog
.sy cat \\$1\" bring in postscript file
...\" the following line matches the tempdict above
\\!end % tempdict %
\\!PE
\\!.
.sp \\$2u	\" move below the image
..
.de pF
.ie     \\*(f1 .ds f1 \\n(.f
.el .ie \\*(f2 .ds f2 \\n(.f
.el .ie \\*(f3 .ds f3 \\n(.f
.el .ie \\*(f4 .ds f4 \\n(.f
.el .tm ? font overflow
.ft \\$1
..
.de fP
.ie     !\\*(f4 \{\
.	ft \\*(f4
.	ds f4\"
'	br \}
.el .ie !\\*(f3 \{\
.	ft \\*(f3
.	ds f3\"
'	br \}
.el .ie !\\*(f2 \{\
.	ft \\*(f2
.	ds f2\"
'	br \}
.el .ie !\\*(f1 \{\
.	ft \\*(f1
.	ds f1\"
'	br \}
.el .tm ? font underflow
..
.ds f1\"
.ds f2\"
.ds f3\"
.ds f4\"
'\" t 
.ta 8n 16n 24n 32n 40n 48n 56n 64n 72n  
.TH "smacq" "1" 
.SH "NAME" 
smacq \(em System for Modular Analysis and Continuous Queries 
.SH "DESCRIPTION" 
.PP 
The System for Modular Analysis and Continuous Queries 
(SMACQ) is an extensible component system for analyzing streams 
of structured data.  This manpage describes the modules 
included with the system by default.  
 
.PP 
These modules are used by SMACQ's command-line utilities 
\fBsmacqp(1)\fP  and \fBsmacqp(1)\fP       as well as a C API described in \fBsmacq-embed(3)\fP. 
 
.SH "STANDARD INPUT/OUTPUT MODULES" 
.SS "Print" 
.PP 
\fBprint\fP [-v]  [-B]  [-d \fIdelimiter\fP] fields \&...  
.PP 
Print the specified fields for every record that has them If 
the \fB-v\fP option is given, then warnings are 
printed when fields aren't present, and field values are 
preceded by the field name.  If the \fB-B\fP option 
is specified, then output is not buffered (output is flushed 
after each record).  Fields are separated by a delimiter 
string which can be specified by \fB-d\fP and 
defaults to TAB. 
 
.SS "Tabular Input" 
.PP 
\fBtabularinput\fP [-d \fIdelimiter\fP]  [-f \fIfilename\fP]  [field [:type]  \&...]  
.PP 
Read records from STDIN by default, or a filename specified with 
the \fB-f\fP option, one per line, with fields delimited 
by TAB or an alterate delimiter specified by 
\fB-d\fP.  If field names are specified, data 
columns are assigned those names sequentially.  If no field 
names are specified, or there are more columns that field 
names, unnamed fields are named numericaly starting at 1.  
 
.PP 
Field types can be specified by appending a colon and the type name to the end of the field name. 
If no type is sepecified for a field, it is treated as a double if possible, or a string otherwise. 
 
.SS "Packet Capture" 
.PP 
\fBpcaplive\fP [-i \fIinterface\fP]  [-s \fIsnaplen\fP]  [-p]  [\fIfilter\fP \&...]  
.PP 
The \fBpcaplive\fP module reads packets from a 
network interfaces using libpcap.  It can only be used at the 
beginning of a pipeline.  Root privileges are typically 
required to run this module. 
 
.PP 
The \fB-i\fP option specifies an interface to 
listen on (default is \fBany\fP).  The 
\fB-s\fP option specifies the maximum number of 
bytes per packet to capture (default is 68).  The 
\fB-p\fP specifies that the interface should NOT be 
placed in promiscuous mode. 
 
.PP 
An optional filter string is a BPF filter string (see 
\fBtcpdump\fP(1)). 
 
.SS "Packet Trace File" 
.PP 
\fBpcapfile\fP \fIfilename\fP \&...  
.PP 
\fBpcapfile\fP -w \fIfilename\fP  [-l ]  [-s \fImegabytes\fP]  
.PP 
The \fBpcapfile\fP module either reads from or 
writes to a tcpdump-style, libpcap packet trace file.  If the 
module is at the beginning of a pipeline, it reads from a 
file.  Otherwise it writes data to a file. 
 
.PP 
When reading, one or more files must be specified.  Use 
\fB-\fP for stdin.  Input files that are 
compressed with \fBgzip\fP are supported 
automatically.  If the \fB-l\fP option 
is specified then files are read from STDIN instead of  
the arguments. 
 
.PP 
When writing, a single output file must be specified with 
\fB-w\fP. 
 
.PP 
The \fB-s\fP option specifies the maximum file size (in 
megabytes) for output files.  If specified, the output file 
will have a two-digit suffix number appended and output will 
be split between as many files as necessary. 
 
.SS "Socket" 
.PP 
\fBsocket\fP [-p \fIport\fP]  [-h \fIhost\fP]  [-d]  
.PP 
The \fBsocket\fP module is used to send records 
across the network to another instantiation of the 
\fBsocket\fP module.  It can be used in two 
different ways: as a producer who receives data from the 
network, or as a consumer that writes data to a network.  If 
the module is at the beginning of a pipeline, it is assumed to 
be a server.  Otherwise it is a consumer that writes data to 
the network. 
 
.PP 
The \fB-h\fP and \fB-p\fP options 
specify a host and port, respectively.  The host option is required 
for a consumer.  The default port is 3000. 
 
.PP 
The \fB-d\fP option is only valid in the server context. 
If specified, the module will continue to accept new 
connections forever and will never exit.  Without this option, 
the server will accept a single connection, process it until 
it closes, and then terminate. 
 
.SH "FILTER MODULES (BOOLEAN FUNCTIONS)" 
.SS "IP Address Mask" 
.PP 
\fBmask\fP\fIfield\fP [!]\fIaddr/cidr\fP  
.PP 
The "addr/cidr" argument is a CIDR netmask.  If the mask size 
is not specified, 32 is assumed.  An object is filtered out if 
and only if the specified field does not exist or does not 
match the given netmask.  If the address begins with a '!', 
then the logic is reversed and the object is filtered out if 
the field does match the netmask. 
 
.SS "Substring" 
.PP 
\fBsubstr\fP [\fIfield\fP] \fIstring\fP  [ ; \fIstring\fP \&...]  
.PP 
Search for each byte string in the specified field, or in the 
whole data object if no field is given.  If multiple strings 
are given, then each string corresponds to an output channel, 
and the object will be output only on the channel(s) that 
match. 
 
.SS "Last" 
.PP 
\fBlast\fP [-t \fItime\fP]  [\fIfields\fP \&...]  
.PP 
If any fields are specified, treat those fields as a tuple and keep track of the last object seen  
with that tuple value.  After there is no more data, output the objet for each tuple value. 
 
.PP 
The \fB-t\fP option specifies, as a real number, 
the number of seconds between periodic updates.  After the specified 
amount of time, the last object seen for each tuple value will be emitted  
(just as is done at the end of the data stream).  At the end of the update, an 
object of type "\fBrefresh\fP" will be sent with a  
"\fBtimeseries\fP" field of type "\fBtimeval\fP" 
containing the time. 
Note: Time is not the wall-clock time, but is instead the 
time stored in the record in the 
"\fBtimeseries\fP" field of type 
"\fBtimeval\fP".  The \fB-t\fP cannot 
be used with records that do not have this field. 
 
.SS "Filter" 
.PP 
\fBfilter\fP \fIfield [[<=>] value] ...\fP \&...  
.PP 
Filter out all objects in the stream that do not satisfy all 
of the specified criteria.  Expressions can be arbitrarily complex and include AND and OR  
statements and parentheses for grouping. 
 
.PP 
This is the select (sigma) 
operation from relational algebra ("where" in SQL). 
 
.SS "Unique Filter" 
.PP 
\fBuniq\fP [-m \fImegabytes\fP]  \fIfields\fP \&...  
.PP 
Treat the specified field(s) as a tuple and filter out all 
occurrences of duplicate values of that tuple. 
 
.PP 
The \fB-m\fP option specifies that a probabilistic 
algorithm using a fixed amount of memory (specified in 
megabytes) should be employed.  Some records may be mistakenly 
filtered, but some large datasets cannot be processed with a 
perfect algorithm. 
 
.SS "Top" 
.PP 
\fBtop\fP [-m \fImegabytes\fP]  [-r \fIdeviation\fP]  \fIfields\fP \&...  
.PP 
Treat the specified field(s) as a tuple and count the number 
of occurrences of each values of that tuple.  Filter out all 
records except those whose occurrence deviates from the 
average by more than a factor of 
\fBdeviation\fR.  If no 
\fB-r\fP option is specified, the default 
deviation threshold is 1. 
 
.PP 
If \fB-m\fP is specified, then probabilistic 
counters are used, consuming a max of 
\fBmegabytes\fR memory, at the expense of 
some records not being filtered even though they're value is 
rare. 
 
.PP 
It is often useful to follow this module with 
\fBuniq\fP in order to get exact counts for all 
records that pass this filter. 
 
.SS "Head" 
.PP 
\fBhead\fP \fInumber\fP  
.PP 
Pass the first \fBnumber\fR records through 
and then end the pipeline.  Those records will be processed by 
all subsequent modules in the pipeline and the program will 
then terminate. 
 
.SH "ANALYSIS MODULES" 
.SS "Constant Annotation" 
.PP 
\fBconst\fP \fIstring\fP  [\fIfield\fP]  
.PP 
Annotate each object with a field containing the specified string constant. 
If a field name is specified, it will be used.  Otherwise, the name will be the same 
as the value string. 
 
.SS "Counter" 
.PP 
\fBcounter\fP [-f \fIcountname\fP]  [\fIfields\fP \&...]  
.PP 
If no fields are specified, simply count the number of records 
seen.  If one or more fields are specified, treat those fields 
as a tuple and count the number of occurrences of each value 
for that tuple.   
 
.PP 
The count value is added to the record as an annotation of type int and name "counter" 
unless the "-f" is used to specify an alternate name for the field. 
 
.SS "Discrete Probability Density Function" 
.PP 
\fBpdf\fP 
.PP 
Assemble a stream of input records with "count" fields. 
When a "refresh" record is received or the data flow ends, then  
use the "count" fields to calculate the fraction of the total  
that each record is responsible.  Attach this value as a "probability" 
field of type "double". 
calculate then use the  
 
.SS "Project" 
.PP 
\fBproject\fP \fIfields\fP \&...  
.PP 
Replace all objects in the input stream with new objects 
containing only the specified fields.  This is the project (Pi) 
operation from relational algebra ("select <fields>" in SQL). 
 
.SS "Rename" 
.PP 
\fBrename\fP \fIoldfield newfield\fP \&...  
.PP 
Given a list of alternating old and new field names, make a copy of the old field with the new name. 
Combined with the Project module, this can implement the rename (rho) 
operation from relational algebra ("as" in SQL). 
 
.SS "Delta" 
.PP 
\fBdelta\fP \fIxfield\fP  
.PP 
For each data object seen, compute the delta from the previous x field to this current xfield. 
The data object is annotate with a "\fBdelta\fP" 
field of type "\fBdouble\fP" containing the result.  The x field must be convertable to doubles as well. 
 
.SS "Derivative" 
.PP 
\fBderivative\fP \fIyfield\fP \fIxfield\fP  
.PP 
For each data object seen, compute the derivative of the y field with respect to the x field 
between this point and the last object seen.  The data object is annotate with a "\fBderivative\fP" 
field of type "\fBdouble\fP" containing the result.  The x and y fields must be convertable 
to doubles as well. 
 
.SS "Entropy" 
.PP 
\fBentropy\fP 
.PP 
This module expects a series of data objects with "\fBprobability\fP" fields 
and computes the Shannon entropy for that series.   
When the data stream ends or a "\fBrefresh\fP" object is seen, it is assumed 
that every ocurring value has been seen and the entropy for the series is calculated 
and added as an annotation of type \fBdouble\fP to a refresh object.   
See the "\fBlast\fP" module for more information on \fBrefresh\fP objects. 
 
.SS "Flow ID" 
.PP 
\fBflowid\fP [-t \fItime\fP]  [-r ]  \fIfields\fP \&...  
.PP 
Treat the specified field(s) as a tuple and assign a unique 
flow id number to each object based on the typle value.  The 
annotated field is called "flowid".  All but the first packet 
will be filtered out. 
 
.PP 
The \fB-r\fP option specifies that the same flow id 
should be assigned to packets in the reverse direction. 
Separate flow statistics will be kept for each direction. 
 
.PP 
The \fB-t\fP option specifies a number of seconds 
idle time before a flow is timed out.  When it times out a 
REFRESH record with the flows identifying fields (as specified 
in the arguments), the current time (timeseries) and the 
packet and byte counters ("packets", "packetsback", "bytes", 
"bytesout") and the "start" and "finish" times. 
 
.SS "Group-By" 
.PP 
\fBgroupby\fP \fIfields\fP \&...  --   \fIquery\fP \&...  
.PP 
Treat the specified field(s) as a tuple and instantiate the 
specified query for each tuple.  If a record of type "refresh" 
is received, then the pipeline for that tuple will be gracefully 
terminated. 
 
.SS "Time Sort" 
.PP 
\fBfifodelay\fP [-t \fItime\fP]  [-i \fIinput-time-field\fP]  [-o \fIoutput-time-field\fP]  
.PP 
Sort a series of input records and output them sorted by an 
output time field that is specified with the 
\fB-o\fP option and defaults to "timeseries".  All records 
that are past the edge time are immediately updated.  The edge 
time is determined by the input time field (specified with the 
\fB-i\fP option and defaullting to "timseries") and 
a time delay which is specified with the \fB-t\fP 	option which defaults to 0 seconds. 
 
.SH "SEE ALSO" 
.PP 
\fBsmacqq\fP(1),  
\fBsmacqp\fP(1),  
\fBdts\fP(3),  
\fBdts-modules\fP(3),  
\fBsmacq-modules\fP(3), 
\fBsmacq-embed\fP(3) 
 
...\" created by instant / docbook-to-man, Wed 19 Mar 2003, 21:14 
