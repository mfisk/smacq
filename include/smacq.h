#ifndef SMACQ_H
#define SMACQ_H

#include <stdint.h>
#include <stdio.h>
#include <sys/time.h>

/* Turn on performance optimizations: */
#ifdef SMACQ_NO_OPT
# define SMACQ_NO_OPT_DTS
# define SMACQ_NO_OPT_DATAFLOW
#endif

#ifndef SMACQ_NO_OPT_DTS
# define SMACQ_OPT_NOPTHREADS
# define SMACQ_OPT_NOMSGS
# define SMACQ_OPT_FORCEFIELDCACHE
# define SMACQ_OPT_RUNRING
# define SMACQ_OPT_DTS_FREELIST
#endif

#ifndef SMACQ_NO_OPT_DATAFLOW
# define SMACQ_OPT_VECTORS
# define SMACQ_OPT_CHILDREN
# define SMACQ_OPT_HEADS
# define SMACQ_OPT_TAILS
#endif

#if defined(WIN32) && !defined(inline)       
	/* Visual C++ uses the keyword "__inline" rather than "inline" */
	#define inline __inline 
#endif

#ifndef SMACQ_OPT_NOPTHREADS
#include <pthread.h>
#else
#define pthread_mutex_init(x,y)
#define pthread_mutex_destroy(x)
#define pthread_mutex_lock(x)
#define pthread_mutex_unlock(x)

#define pthread_cond_init(x,y)
#define pthread_cond_destroy(x)
#define pthread_cond_wait(x,y)
#define pthread_cond_broadcast(x)
#define _PTHREAD_H
#endif

#ifndef BEGIN_C_DECLS
# ifdef __cplusplus
#   define BEGIN_C_DECLS extern "C" {
#   define END_C_DECLS }
# else
#   define BEGIN_C_DECLS
#   define END_C_DECLS
# endif
#endif

enum _smacq_result { SMACQ_FREE=1, SMACQ_PASS=2, SMACQ_ERROR=4, SMACQ_END=8, SMACQ_CANPRODUCE=256, SMACQ_PRODUCE=512};
typedef enum _smacq_result smacq_result;

#define SMACQ_MULTITHREAD 65536

struct smacq_init;
struct arglist;
struct state; // Private to modules
struct _smacq_module;
typedef struct _smacq_module smacq_graph;

#include "util.c"
#include "darray.c"
#include "smacq_args.h"
#include "dts.h"

typedef struct _smacq_env {
  dts_environment * types;
} smacq_environment;

typedef smacq_result smacq_thread_fn(struct smacq_init *);

struct smacq_init {
  int isfirst;
  int islast;
  char ** argv;
  int argc;
  smacq_environment * env;
  void * state;
  smacq_graph * self;
  smacq_thread_fn * thread_fn;
};

/*
 * Interrogation structures
 */
struct smacq_module_algebra {
  unsigned int vector:1;
  unsigned int boolean:1;
  unsigned int demux:1;
  unsigned int nesting:1;
};

typedef smacq_result smacq_produce_fn(struct state * state, const dts_object **, int * outchan);
typedef smacq_result smacq_consume_fn(struct state * state, const dts_object *, int * outchan);
typedef smacq_result smacq_init_fn(struct smacq_init *);
typedef smacq_result smacq_shutdown_fn(struct state *);
#include <SmacqModule.h>
typedef SmacqModule*  smacq_constructor_fn(struct smacq_init *);

struct smacq_functions {
  smacq_produce_fn * produce;
  smacq_consume_fn * consume;
  smacq_init_fn * init;
  smacq_shutdown_fn * shutdown;
  smacq_thread_fn * thread;
  smacq_constructor_fn * constructor;
  struct smacq_module_algebra algebra;
};

BEGIN_C_DECLS

/*
 * Module threading routines 
 */
smacq_result smacq_thread_init(struct smacq_init * volatile_context);
smacq_result smacq_thread_consume(struct state * state, const dts_object * datum, int * outchan);
smacq_result smacq_thread_produce(struct state * state, const dts_object ** datum, int *outchan);
smacq_result smacq_thread_shutdown(struct state * state);

#define SMACQ_THREADED_MODULE(THREAD) { \
	init: smacq_thread_init, \
	consume: smacq_thread_consume, \
	produce: smacq_thread_produce, \
	shutdown: smacq_thread_shutdown, \
	thread: THREAD };

const dts_object * smacq_read(struct smacq_init * context);
void smacq_write(struct state * state, dts_object * datum, int outchan);
void smacq_decision(struct smacq_init * context, const dts_object * datum, smacq_result result);
int smacq_flush(struct smacq_init * context);


/*
 * User Interface to main system
 */
enum smacq_scheduler { ITERATIVE, RECURSIVE, LOOP, THREADED };
int smacq_start(smacq_graph *, enum smacq_scheduler, dts_environment *);
void smacq_init_modules(smacq_graph *, smacq_environment *);
smacq_graph * smacq_build_pipeline(int argc, char ** argv);
smacq_graph * smacq_build_query(dts_environment * tenv, int argc, char ** argv);
int smacq_execute_query(int argc, char ** argv);
smacq_graph * smacq_add_new_child(smacq_graph * parent, int argc, char ** argv);
int smacq_add_child(smacq_graph * parent, smacq_graph * child);
void smacq_replace_child(smacq_graph * parent, int num, smacq_graph * newchild);
void smacq_add_parent(smacq_graph * child, smacq_graph * parent);
void smacq_remove_child(smacq_graph * parent, int childnum);
void smacq_remove_parent(smacq_graph * child, const smacq_graph * parent);
smacq_graph * smacq_merge_graphs(smacq_graph *);
smacq_graph * smacq_graph_add_graph(smacq_graph * a, smacq_graph * b);
smacq_graph * smacq_graph_clone(smacq_environment *, smacq_graph *);

smacq_graph * smacq_new_module(int argc, char ** argv);
void smacq_free_module(smacq_graph * f);
void smacq_destroy_graph(smacq_graph * f);
smacq_graph * smacq_clone_child(smacq_graph * parent, int child);
smacq_graph * smacq_clone_tree(smacq_graph * donorParent, smacq_graph * newParent, int child);

int smacq_graph_print(FILE * fh, smacq_graph * f, int indent);
int smacq_graphs_print(FILE * fh, smacq_graph * f, int indent);
double smacq_graph_count_nodes(smacq_graph * f);

END_C_DECLS

enum smacq_log_level { INFO, WARN, ERROR, };

static inline void smacq_log(char * name, enum smacq_log_level level, char * msg) {
  fprintf(stderr, "%s: %s\n", name, msg);
}

#include "smacq-internal.h"
#include "types-inline.c"


#endif

