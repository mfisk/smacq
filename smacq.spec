Summary: SMACQ analysis tool for streaming data
Name: smacq
Version: 2.2
Release: 1
License: GPL
Group: Applications/Internet
URL: http://smacq.sourceforge.net/
Source0: %{name}-%{version}.tar.gz
BuildRoot: %{_tmppath}/%{name}-%{version}-%{release}-root
Requires: libgda, glib2, zlib, libpcap
BuildRequires: gcc-c++, libgda-devel, glib2-devel, zlib-devel

%description

The System for Modular Analysis and Continuous Queries (SMACQ) is a
modular platform for analyzing and querying large datasets, including
streaming network data, using features from databases, UNIX pipelines,
and modular intrusion detection systems. 

%prep
%setup -q


%build

%configure --with-docdir=/usr/share/doc/${name}-${version}.


%install
rm -rf $RPM_BUILD_ROOT
make install DESTDIR=${RPM_BUILD_ROOT}

%clean
rm -rf $RPM_BUILD_ROOT

%files
%defattr(-,root,root,-)
%attr(0755,root,root) /usr/bin/*
%attr(0755,root,root) /usr/share/man/man1/*
%attr(0755,root,root) /usr/share/man/man3/*
%attr(0755,root,root) /usr/share/doc/smacq-${version}/*

#%doc AUTHORS README TODO doc/smacq.1.pdf doc/smacqq.1.pdf doc/gotsmacq.png doc/icon.png doc/index.html doc/api/refman.pdf 

%changelog
* Tue Mar 22 2005 Stephen J Smoogen <smoogen@smoogen3.lanl.gov> - 
- Initial build.

