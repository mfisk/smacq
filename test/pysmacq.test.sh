#! /bin/sh

PYTHONPATH=$OUTDIR/../pysmacq/.libs:$SRCDIR/../pysmacq:$TEST
export PYTHONPATH

TESTFILE=0325@1112-snort.pcap
export TESTFILE

python2.3 <<EOF
import smacq
print "Imported successfully"

pcapfile = '0325@1112-snort.pcap'

q = smacq.SmacqQuery('pcapfile(%s)' % (pcapfile))
q >>= 'where srcport > 137'

print q

for d in q:
	print d['srcip'], d['dstip'], d['srcport'], d['dstport'], d.has_key('foo')

EOF

