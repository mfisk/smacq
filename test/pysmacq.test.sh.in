#! /bin/sh

PYTHONPATH=$OUTDIR/../pysmacq/.libs:$SRCDIR/../pysmacq:$TEST
export PYTHONPATH

TESTFILE=0325@1112-snort.pcap
export TESTFILE

@PYTHON@ <<EOF
import smacq
print "Imported successfully"

pcapfile = '0325@1112-snort.pcap'

q = smacq.SmacqQuery('pcapfile(%s)' % (pcapfile))
q >>= 'where srcport > 137'

print 'Query string:', q
for d in q:
	print repr(d), d

EOF

