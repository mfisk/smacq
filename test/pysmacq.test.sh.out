Imported successfully
filter '(srcport > "137")' from (pcapfile '0325@1112-snort.pcap' )
192.168.0.54 192.168.0.210 1043 139 False
192.168.0.54 192.168.0.210 1043 139 False
192.168.0.54 192.168.0.210 1043 139 False
192.168.0.54 192.168.0.210 1043 139 False
192.168.0.54 192.168.0.200 1051 445 False
192.168.0.54 192.168.0.200 1052 139 False
192.168.0.54 192.168.0.200 1052 139 False
192.168.0.54 192.168.0.200 1052 139 False
192.168.0.54 192.168.0.200 1052 139 False
192.168.0.54 192.168.0.200 1052 139 False
192.168.0.54 192.168.0.200 1053 21 False
192.168.0.54 192.168.0.200 1054 80 False
192.168.0.54 192.168.0.200 1055 25 False
192.168.0.54 192.168.0.200 1056 110 False
192.168.0.54 192.168.0.200 1057 111 False
192.168.0.54 192.168.0.200 1058 79 False
192.168.0.54 192.168.0.200 1059 53 False
192.168.0.54 192.168.0.200 1053 21 False
192.168.0.54 192.168.0.200 1054 80 False
192.168.0.54 192.168.0.200 1057 111 False
192.168.0.54 192.168.0.200 1058 79 False
192.168.0.54 192.168.0.200 1059 53 False
192.168.0.54 192.168.0.200 1055 25 False
192.168.0.54 192.168.0.200 1056 110 False
192.168.0.54 192.168.0.200 1053 21 False
192.168.0.54 192.168.0.200 1054 80 False
192.168.0.54 192.168.0.200 1057 111 False
192.168.0.54 192.168.0.200 1058 79 False
192.168.0.54 192.168.0.200 1059 53 False
192.168.0.54 192.168.0.200 1055 25 False
192.168.0.54 192.168.0.200 1056 110 False
192.168.0.54 192.168.0.200 1052 139 False
192.168.0.54 192.168.0.200 1052 139 False
192.168.0.54 192.168.0.200 1060 445 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1062 135 False
192.168.0.54 192.168.0.200 1062 135 False
192.168.0.54 192.168.0.200 1062 135 False
192.168.0.54 192.168.0.200 1062 135 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1062 135 False
192.168.0.54 192.168.0.200 1062 135 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
192.168.0.54 192.168.0.200 1061 139 False
